### ISSUE TITLE
terragrunt conflicts with tfenv

### ISSUE THREAD

**Issue by MacFlurry** [gruntwork-io/terragrunt#580]
Hello,

OS macOS 10.14
trying to install terragrunt

`$ brew install terragrunt`

> Updating Homebrew...
> ==> Auto-updated Homebrew!
> Updated 1 tap (homebrew/core).
> No changes to formulae.
> 
> ==> Installing dependencies for terragrunt: terraform
> ==> Installing terragrunt dependency: terraform
> Error: Cannot install terraform because conflicting formulae are installed.
>   tfenv: because tfenv symlinks terraform binaries

> Please `brew unlink tfenv` before continuing.

> Unlinking removes a formula's symlinks from /usr/local. You can
> link the formula again after the install finishes. You can --force this
> install, but the build may fail or cause obscure side-effects in the
> resulting software.

So terragrunt won't install because of Tfenv exists.
Any help ?

**Comment by brikis98** [comment]
The Homebrew installer for Terragrunt is maintained by the community, so to be honest, I'm not sure what is the right fix for this. Suggestions and PRs are very welcome!

**Comment by udondan** [comment]
You can just skip dependencies when installing:

```
brew install --ignore-dependencies terragrunt
```



Or in a [Brewfile (Homebrew Bundle)](https://github.com/Homebrew/homebrew-bundle):

```
brew "awscli"
brew "aws-okta"
brew "terraform_landscape"
brew "terragrunt", args: ["ignore-dependencies"]
brew "tfenv"
```

Then it can be installed via `brew bundle`.

But I'd also prefer terraform not to be a hard dependency. Though I couldn't find the source for the Homebrew config to propose a change.

**Comment by evanstoddard23** [comment]
An issue with using a Brewfile with `ignore-dependencies` is that flag is ignored on upgrades, giving the same linking issues. Since homebrew core doesn't support options which would allow terraform install to be optional, I'm considering a PR to remove terraform as a dependency and adding info about installing terraform/tfenv in the caveats section. Curious what others, especially maintainers might think about this.

**Comment by blairham** [comment]
The real fix is to just remove the dependency, as terragrunt does not require it to actually execute, it's an external dependency. Anyone using terragrunt, already knows they will need terraform, as it will also give you an error if you run it without terraform installed.

**Comment by robmorgan** [comment]
I believe this can be closed now that https://github.com/Homebrew/homebrew-core/pull/73326 is merged.

**Comment by mmindenhall** [comment]
I'm going to submit a new PR that removes `terraform` as a dependency, and adds the following `caveats` section.  Please chime in with any additions or edits you think might be helpful!

```ruby
  def caveats
    <<~EOS
      Terragrunt has been installed as
        #{HOMEBREW_PREFIX}/bin/terragrunt

      Terragrunt requires a version of `terraform` to be in the user's path.

      Teams using terragrunt/terraform need to use the same version of
      terraform. To prevent accidental version changes via `brew upgrade`, we
      recommend using `tfenv` to install and manage `terraform` versions:
        brew install tfenv
        tfenv install <required terraform version>
        tfenv use <required terraform version>

      See: https://github.com/tfutils/tfenv

      Terraform can also be installed directly with homebrew.  This is only
      recommended when NOT working as part of a team (i.e., for solo projects,
      learning, or experimenting):
        brew install terraform
    EOS
  end
```

**Comment by mmindenhall** [comment]
No joy!  In the first PR it sounded like homebrew was amenable to the removal of terraform as a dependency as long as there were clear instructions to the user.  In the 2nd PR, they made it clear that even with clear instructions for how to install terraform separately, they will not allow removal of terraform as a dependency in homebrew-core, because then the formula may not work when installed.  They suggested creating a separate homebrew tap, then the formula can be managed however we want.  If someone at gruntwork wants to take this up:

1. Create a [new tap](https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap) repo under the `gruntwork-io` account.  If there are other gruntwork tools that might be included in the future, it would probably make sense to name the repo `homebrew-gruntwork` instead of `homebrew-terragrunt`.
2. Add the terragrunt formula (similar to what was done in https://github.com/Homebrew/homebrew-core/pull/75836)
3. Create a PR against homebrew-core to [remove](https://docs.brew.sh/Deprecating-Disabling-and-Removing-Formulae) the existing formula
4. Update the [docs](https://terragrunt.gruntwork.io/docs/getting-started/install/) with instructions to install using the new tap.

**Comment by mmindenhall** [comment]
I was able to implement a workaround for this using a git `post-merge` hook.  Here's how to do it:

1. `cd $(brew --repo homebrew/core)` to get to the repo root.
1. From there, create a file under `.git/hooks` called `post-merge`.  The file should contain the following (if you already have a `post-merge` file for some reason, add the code after the first line):

    ```
    #!/usr/bin/env bash
    
    changed_files="$(git diff-tree -r --name-only --no-commit-id ORIG_HEAD HEAD)"
    
    check_run() {
    	echo "$changed_files" | grep --quiet "$1" && eval "$2"
    }
    
    # Remove terraform dependency from terragrunt
    check_run terragrunt.rb "sed -i '' 's/depends_on \"terraform\"//g' Formula/terragrunt.rb"
    ```
1. Make sure the file is executable:  `chmod +x .git/hooks/post-merge`

Note that if you use `gnu-sed` (installed as `gsed`) instead of `sed` native in macOS, you'll need to modify the last line slightly:

```
check_run terragrunt.rb "gsed -i 's/depends_on \"terraform\"//g' Formula/terragrunt.rb"
```

With this hook, whenever you run `brew update` and `terragrunt.rb` is included in the update, the dependency on `terraform` will be removed from the updated file.  Then when you run `brew upgrade`, the upgrade will succeed without having `terraform` installed via homebrew.

I was worried that having a modified version of the `terragrunt` formula would cause an error the next time `brew update` ran with a terraform change, but there was just an update today that succeeded, so I think this works.

**Comment by jav-12** [comment]
So at this point what is the best way to avoid the terraform dependency in Terragrunt. This is very annoying every time Terragrunt needs to be updated and the solution given above didn't work for me.

**Comment by lorengordon** [comment]
Based on the response from the Homebrew folks, sounds like just don't use brew to install terragrunt.

> So at this point what is the best way to avoid the terraform dependency in Terragrunt. This is very annoying every time Terragrunt needs to be updated and the solution given above didn't work for me.

**Comment by mmindenhall** [comment]
@jav-12, what's your dev environment (mac, windows, linux)?

**Comment by jav-12** [comment]
Ups sorry. I'm using Mac BigSur @mmindenhall

**Comment by mmindenhall** [comment]
Hi, this approach is working for me and I assume (from the 🚀 responses) at least a few others.  My first suggestion is to double-check that you did everything above correctly:

1. You created the `post-merge` file in the `.git` folder within the `homebrew-core` folder (mine is under `/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core`.
2. Ensure that the `post-merge` file has execute permissions.
3. Make sure that you're using the version of `sed` that comes pre-installed in MacOS (running the command `which sed` should return `/usr/bin/sed`).  If you have installed another version of `sed` (e.g., gnu sed from homebrew) you'll need to change that last line to make the intended change with your version of `sed`.

If that still doesn't work, you could write a small script that will manually run the `sed` command against the `terragrunt.rb` file.  Run the command manually before the next time you run `brew upgrade`.  Then run the command after any `brew upgrade` that includes a new version of terragrunt.  You could also create an alias to do this automatically after every upgrade so you don't have to think about it:

```sh
alias brewup='brew upgrade && ~/scripts/fix-terragrunt.sh'
```

Hope that helps!

**Comment by jorhett** [comment]
I've made a PR trying the options used here https://stackoverflow.com/questions/21115259/alternative-dependencies-in-a-homebrew-formula-e-g-for-use-with-gcc

If they won't accept this, perhaps we should add a tap to this repo?

**Comment by blairham** [comment]
I'm very sorry to necro this. A co-worker brought me back to this. I've resolved this using [asdf](https://asdf-vm.com/) instead of Homebrew to install Terragrunt and Terraform.

**Comment by yhakbar** [comment]
Hey folks. This issue has become very stale, and [tool managers](https://terragrunt.gruntwork.io/docs/getting-started/install/#install-via-tool-manager) are able to handle dynamically switching versions of Terragrunt/OpenTofu/Terraform without issue. Folks are also [installing from Homebrew](https://terragrunt.gruntwork.io/docs/getting-started/install/#install-via-a-package-manager) without issue.

I'm not sure any issue still exist here, so I'm going to be closing it preemptively. If we can reliably reproduce this issue, please comment that you're still blocked by it, and we can reopen the issue to help you get unblocked and install Terragrunt.
